Adding kfdef add and delete

jbusche opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
This PR applies and deletes the  kfdef issues https://github.com/opendatahub-io/distributed-workloads/issues/16 and https://github.com/opendatahub-io/distributed-workloads/issues/19

It requires the CodeFlare Operator to already be installed, (which Kevin is taking care of in his PR #22

Then, my portion of the E2E test (section install_distributed_workloads_kfdef) does the following:
1.  Installs the kfdef
2. Makes sure the 3 expected pods (mcad-controller, instascale and kuberay operator) are running.  (NOTE - Once the KFDEF has Instascale as optional, or if we have two KFDEFs, then this test might need to be slighly modified to account for that.)
3. It makes sure the codeflare-nodebook imagestream is there

Then, my portion of the E2E test (section uninstall_distributed_workloads_kfdef) does the following:
1. Uninstalls the kfdef. (Note, this can take up to 300+ seconds, depending on where in the reconcile loop the MCAD controller is at.
2. Checks to make sure all three pods are gone.
3. It makes sure the codeflare-nodebook imagestream is removed.

## How Has This Been Tested?
I have been modifying and running the E2E test with peak/run.sh.  I'll pull fresh my PR and post the test results below before I take the PR out of draft form.

## Merge criteria:
<!--- This PR will be merged by any repository approver when it meets all the points in the checklist -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

- [ ] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

jbusche commented:
I think it's looking good - Here's the output I'm seeing: (Note, I don't like this ray.sh test, will create an issue, see if we can remove it or comment it out.)
```
./run.sh 
++++++ /root/JIM/peak/operator-tests/opendatahub-kubeflow/ray/tests/basictests
Now using project "basictests-bafq" on server "https://api.jimbig412.cp.fyre.ibm.com:6443".

++++ /root/JIM/peak/operator-tests/opendatahub-kubeflow/ray/tests/basictests/ray.sh
./run.sh: line 124: /root/JIM/peak/operator-tests/opendatahub-kubeflow/ray/tests/basictests/ray.sh: Permission denied
failed: /root/JIM/peak/operator-tests/opendatahub-kubeflow/ray/tests/basictests/ray.sh
++++++ /root/JIM/peak/operator-tests/opendatahub-kubeflow/tests/basictests
Already on project "basictests-bafq" on server "https://api.jimbig412.cp.fyre.ibm.com:6443".

++++ /root/JIM/peak/operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh

Running example test

Running operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:15: executing 'oc project opendatahub' expecting success...


✔ SUCCESS after 0.117s: operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:15: executing 'oc project opendatahub' expecting success

Running operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:16: executing 'oc get pods' expecting success...


✔ SUCCESS after 0.096s: operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:16: executing 'oc get pods' expecting success


Installing Codeflare Operator


Installing distributed workloads kfdef

Running operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:25: executing 'oc apply -f https://raw.githubusercontent.com/opendatahub-io/distributed-workloads/main/codeflare-stack-kfdef.yaml -n opendatahub' expecting success...


✔ SUCCESS after 2.989s: operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:25: executing 'oc apply -f https://raw.githubusercontent.com/opendatahub-io/distributed-workloads/main/codeflare-stack-kfdef.yaml -n opendatahub' expecting success

Running operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:28: executing 'oc get pod -n opendatahub |grep mcad-controller | awk '{print $2}'' expecting any result and text '1/1'; re-trying every 10s until completion or 1200.000s...


✔ SUCCESS after 10.238s: operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:28: executing 'oc get pod -n opendatahub |grep mcad-controller | awk '{print $2}'' expecting any result and text '1/1'; re-trying every 10s until completion or 1200.000s

Running operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:29: executing 'oc get pod -n opendatahub |grep instascale-instascale | awk '{print $2}'' expecting any result and text '1/1'; re-trying every 10s until completion or 1200.000s...


✔ SUCCESS after 0.118s: operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:29: executing 'oc get pod -n opendatahub |grep instascale-instascale | awk '{print $2}'' expecting any result and text '1/1'; re-trying every 10s until completion or 1200.000s

Running operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:30: executing 'oc get pod -n opendatahub |grep kuberay-operator | awk '{print $2}'' expecting any result and text '1/1'; re-trying every 10s until completion or 1200.000s...


✔ SUCCESS after 10.258s: operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:30: executing 'oc get pod -n opendatahub |grep kuberay-operator | awk '{print $2}'' expecting any result and text '1/1'; re-trying every 10s until completion or 1200.000s

Running operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:33: executing 'oc get imagestreams -n opendatahub codeflare-notebook --no-headers=true |awk '{print $1}'' expecting success and text 'codeflare-notebook'...


✔ SUCCESS after 0.114s: operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:33: executing 'oc get imagestreams -n opendatahub codeflare-notebook --no-headers=true |awk '{print $1}'' expecting success and text 'codeflare-notebook'


Testing MCAD TorchX Functionality


Testing MCAD Ray Functionality


Uninstalling distributed workloads kfdef

NOTE, kfdef deletion can take up to 5-8 minutes...
Running operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:47: executing 'oc delete kfdef codeflare-stack -n opendatahub' expecting success; re-trying every 10s until completion or 1200.000s...


✔ SUCCESS after 336.067s: operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:47: executing 'oc delete kfdef codeflare-stack -n opendatahub' expecting success; re-trying every 10s until completion or 1200.000s

Running operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:50: executing 'oc get pod -n opendatahub -l app=mcad-mcad' expecting any result and text 'No resources found in opendatahub namespace.'; re-trying every 10s until completion or 1200.000s...


✔ SUCCESS after 10.233s: operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:50: executing 'oc get pod -n opendatahub -l app=mcad-mcad' expecting any result and text 'No resources found in opendatahub namespace.'; re-trying every 10s until completion or 1200.000s

Running operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:51: executing 'oc get pod -n opendatahub -l app=instascale-instascale' expecting any result and text 'No resources found in opendatahub namespace.'; re-trying every 10s until completion or 1200.000s...


✔ SUCCESS after 0.138s: operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:51: executing 'oc get pod -n opendatahub -l app=instascale-instascale' expecting any result and text 'No resources found in opendatahub namespace.'; re-trying every 10s until completion or 1200.000s

Running operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:52: executing 'oc get pod -n opendatahub -l app.kubernetes.io/component=kuberay-operator' expecting any result and text 'No resources found in opendatahub namespace.'; re-trying every 10s until completion or 1200.000s...


✔ SUCCESS after 0.102s: operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:52: executing 'oc get pod -n opendatahub -l app.kubernetes.io/component=kuberay-operator' expecting any result and text 'No resources found in opendatahub namespace.'; re-trying every 10s until completion or 1200.000s

Running operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:55: executing 'oc get imagestreams -n opendatahub codeflare-notebook' expecting failure...


✔ SUCCESS after 0.143s: operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:55: executing 'oc get imagestreams -n opendatahub codeflare-notebook' expecting failure


Uninstalling Codeflare Operator

Already on project "opendatahub" on server "https://api.jimbig412.cp.fyre.ibm.com:6443".
One or more tests failed:

	/root/JIM/peak/operator-tests/opendatahub-kubeflow/ray/tests/basictests/ray.sh

./run.sh took 373 seconds
[INFO] Exiting with 1
```

MichaelClifford commented on tests/basictests/distributed-workloads.sh:
```diff
@@ -21,7 +21,16 @@ function install_codeflare_operator() {
 function install_distributed_workloads_kfdef(){
-    header "Installing distributed workloads kfdef"   
+    header "Installing distributed workloads kfdef"
+    os::cmd::expect_success "oc apply -f https://raw.githubusercontent.com/opendatahub-io/distributed-workloads/main/codeflare-stack-kfdef.yaml -n ${ODHPROJECT}"
```
I think this file should come from the repo, not github. 

`oc apply -f ../../../codeflare-stack-kfdef.yaml`

jbusche replied:
OK, sure... let me try it, test it out - thanks!

KPostOffice replied:
I agree, I think it will have to be `{$MY_DIR}/../../codeflare-stack-kfdef.yaml` in order to avoid relative path issues.

jbusche replied:
The file doesn't exist (yet) anyway... should I create the file, store it under resources/codeflare-stack-kfdef.yaml and point to it with:  
```"oc apply -f $RESOURCEDIR/codeflare-stack-kfdef.yaml -n ${ODHPROJECT}"```

jbusche replied:
Oh, wait - I was grepping, not finding... ugh, it is there...  hold on... let's see if this works.

jbusche replied:
This succeeds...  Thanks Kevin! (It didn't like the {}...)  let me get it in the PR:
`"oc apply -f $MY_DIR/../../codeflare-stack-kfdef.yaml -n ${ODHPROJECT}"`

MichaelClifford commented on tests/basictests/distributed-workloads.sh:
```diff
@@ -34,13 +43,28 @@ function tests_mcad_ray_functionality() {
 }
 
 
+function uninstall_codeflare_operator() {
```
Is this second "uninstall_codeflare_operator()" required?

jbusche replied:
Oh shoot, I'm not sure how that got there... will remove.  Thanks!

MichaelClifford reviewed (changes requested):
:+1:  Just  a couple of small comments

jbusche commented:
OK @MichaelClifford I've cleaned up the extra "uninstall_codeflare_operator()" and changed the kfdef to local... (thanks @KPostOffice for the pointer.). Running a test now, will post results soon.

jbusche commented:
Looks good:
```
./run.sh distributed-workloads.sh
by in [INFO] No regex /root/JIM/peak/operator-tests/opendatahub-kubeflow/ray/tests/basictests selected tests were
++++++ /root/JIM/peak/operator-tests/opendatahub-kubeflow/tests/basictests
Now using project "basictests-efcv" on server "https://api.jim412.cp.fyre.ibm.com:6443".

++++ /root/JIM/peak/operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh

Running example test

Running operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:15: executing 'oc project opendatahub' expecting success...


✔ SUCCESS after 0.156s: operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:15: executing 'oc project opendatahub' expecting success

Running operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:16: executing 'oc get pods' expecting success...


✔ SUCCESS after 0.195s: operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:16: executing 'oc get pods' expecting success


Installing Codeflare Operator


Installing distributed workloads kfdef

Running operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:25: executing 'oc apply -f /root/JIM/peak/operator-tests/opendatahub-kubeflow/tests/basictests/../../codeflare-stack-kfdef.yaml -n opendatahub' expecting success...


✔ SUCCESS after 0.498s: operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:25: executing 'oc apply -f /root/JIM/peak/operator-tests/opendatahub-kubeflow/tests/basictests/../../codeflare-stack-kfdef.yaml -n opendatahub' expecting success

Running operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:28: executing 'oc get pod -n opendatahub |grep mcad-controller | awk '{print $2}'' expecting any result and text '1/1'; re-trying every 10s until completion or 1200.000s...


✔ SUCCESS after 316.599s: operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:28: executing 'oc get pod -n opendatahub |grep mcad-controller | awk '{print $2}'' expecting any result and text '1/1'; re-trying every 10s until completion or 1200.000s

Running operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:29: executing 'oc get pod -n opendatahub |grep instascale-instascale | awk '{print $2}'' expecting any result and text '1/1'; re-trying every 10s until completion or 1200.000s...


✔ SUCCESS after 0.208s: operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:29: executing 'oc get pod -n opendatahub |grep instascale-instascale | awk '{print $2}'' expecting any result and text '1/1'; re-trying every 10s until completion or 1200.000s

Running operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:30: executing 'oc get pod -n opendatahub |grep kuberay-operator | awk '{print $2}'' expecting any result and text '1/1'; re-trying every 10s until completion or 1200.000s...


✔ SUCCESS after 10.402s: operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:30: executing 'oc get pod -n opendatahub |grep kuberay-operator | awk '{print $2}'' expecting any result and text '1/1'; re-trying every 10s until completion or 1200.000s

Running operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:33: executing 'oc get imagestreams -n opendatahub codeflare-notebook --no-headers=true |awk '{print $1}'' expecting success and text 'codeflare-notebook'...


✔ SUCCESS after 0.168s: operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:33: executing 'oc get imagestreams -n opendatahub codeflare-notebook --no-headers=true |awk '{print $1}'' expecting success and text 'codeflare-notebook'


Testing MCAD TorchX Functionality


Testing MCAD Ray Functionality


Uninstalling distributed workloads kfdef

NOTE, kfdef deletion can take up to 5-8 minutes...
Running operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:47: executing 'oc delete kfdef codeflare-stack -n opendatahub' expecting success; re-trying every 10s until completion or 1200.000s...


✔ SUCCESS after 643.432s: operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:47: executing 'oc delete kfdef codeflare-stack -n opendatahub' expecting success; re-trying every 10s until completion or 1200.000s

Running operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:50: executing 'oc get pod -n opendatahub -l app=mcad-mcad' expecting any result and text 'No resources found in opendatahub namespace.'; re-trying every 10s until completion or 1200.000s...


✔ SUCCESS after 10.336s: operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:50: executing 'oc get pod -n opendatahub -l app=mcad-mcad' expecting any result and text 'No resources found in opendatahub namespace.'; re-trying every 10s until completion or 1200.000s

Running operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:51: executing 'oc get pod -n opendatahub -l app=instascale-instascale' expecting any result and text 'No resources found in opendatahub namespace.'; re-trying every 10s until completion or 1200.000s...


✔ SUCCESS after 0.136s: operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:51: executing 'oc get pod -n opendatahub -l app=instascale-instascale' expecting any result and text 'No resources found in opendatahub namespace.'; re-trying every 10s until completion or 1200.000s

Running operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:52: executing 'oc get pod -n opendatahub -l app.kubernetes.io/component=kuberay-operator' expecting any result and text 'No resources found in opendatahub namespace.'; re-trying every 10s until completion or 1200.000s...


✔ SUCCESS after 0.158s: operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:52: executing 'oc get pod -n opendatahub -l app.kubernetes.io/component=kuberay-operator' expecting any result and text 'No resources found in opendatahub namespace.'; re-trying every 10s until completion or 1200.000s

Running operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:55: executing 'oc get imagestreams -n opendatahub codeflare-notebook' expecting failure...


✔ SUCCESS after 0.124s: operator-tests/opendatahub-kubeflow/tests/basictests/distributed-workloads.sh:55: executing 'oc get imagestreams -n opendatahub codeflare-notebook' expecting failure


Uninstalling Codeflare Operator

Already on project "opendatahub" on server "https://api.jim412.cp.fyre.ibm.com:6443".
./run.sh took 984 seconds
[INFO] Exiting with 0
```

MichaelClifford commented:
LGTM @jbusche 
Can you rebase and we can merge it :)
